Simplify atc/worker package and extract runtime abstractions

aoldershaw opened the pull request:
<!--
Hi there! Thanks for submitting a pull request to Concourse!

The title of your pull request will be used to generate the release notes.
Please provide a brief sentence that describes the PR, using the [imperative
mood]. Please refrain from adding prefixes like 'feature:', and don't include a
period at the end.

Examples: "Add feature to doohickey", "Fix panic during spline reticulation"

We will edit the title if needed so don't worry about getting it perfect!

To help us review your PR, please fill in the following information.
-->

[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

## What does this PR accomplish?
<!--
Choose all that apply.
Also, mention the linked issue here.
This will magically close the issue once the PR is merged.
-->

The current atc/worker package is extremely fragmented with a ton of indirection and internal interfaces, presumably for the sake of easier testing. Here's a (loosely defined) dependency graph for the existing atc/worker package I put together when trying to understand how things currently interact:

![worker_dep_graph](https://user-images.githubusercontent.com/26583442/109099196-ae187880-76f0-11eb-9501-99084bc1c5e7.png)

The tests all test against these internal interfaces against auto-generated fakes and break upon any sort of code restructuring.

This PR aims to tear out the existing atc/worker package and replace it with the same logic, but with much less indirection - pretty much all abstraction (besides the firm interfaces - Garden, Baggageclaim, and the DB) is limited to method calls. It brings a new test-suite that tests at a higher level against the interfaces that Garden-based workers actually depend on - Garden, Baggageclaim, and the DB - using hand-written in-memory fakes (or, in the case of the DB, an actual Postgres server).

It also introduces tries to come up with some more clean interfaces for implementing other runtimes (esp. k8s) and puts these types in the `atc/runtime` package.

Below is the new model of the world:
![worker_dep_graph (1)](https://user-images.githubusercontent.com/26583442/111242281-3d6cc980-85d5-11eb-9142-fa9aebd7b7d4.png)

In particular, the big box in the first diagram has become the two packages `atc/worker` and `atc/worker/gardenruntime` above.

## Changes proposed by this PR:
<!--
Tell the reviewer What changed, Why, and How were you able to accomplish that?
-->

Still a wip, see the individual commits for much more detailed commit messages

## Notes to reviewer:
<!--
Leave a message to whoever is going to review this PR.
Mainly, pointers to review the PR, and how they can test it.
-->

## Release Note
* We will now error when a suitable worker does not exist rather than waiting forever.

## Contributor Checklist
<!--
Most of the PRs should have the following added to them,
this doesn't apply to all PRs, so it is helpful to tell us what you did.
-->
- [ ] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [ ] [Signed] all commits
- [ ] Added tests (Unit and/or Integration)
- [ ] Updated [Documentation]
- [ ] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist
<!--
This section is intended for the reviewers only, to track review
progress.
-->
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

aoldershaw committed:
re-implement container creation in a simpler way

the existing atc/worker package orchestrates between the DB and
garden+baggageclaim on workers to construct containers+volumes. it
provides a `worker.Pool` for selecting a worker given a strategy, and a
`worker.Client` for running the core steps (check, get, put, task)

the current implementation has a ton of abstraction, presumably for the
sake of easier testing, but at the cost of simplicity. the tests are
also very tightly coupled to the set of abstractions, making it very
difficult to refactor - it was easier to just rewrite it in a new
package, currently atc/worker2, which no code calls at the moment. this
code is very important to be able to refactor as we work on a kubernetes
runtime

currently, I've implemented worker.FindOrCreateContainer, whose
underlying logic was pretty much copied from atc/worker, albeit without
any internal interfaces - it only depends on a garden, baggageclaim, and
the DB (as well as the pool of workers for volume streaming). I've also
implemented a new general interface for running/attaching to processes,
intended to replace the runtime.Runner (which has an interface that only
works for resources). I've also implemented a basic form of worker.Pool
that's only able to lookup volumes. tests to follow in subsequent
commits

what's changed from the original implementation:

structural
----------
* introduce more general runtime types with the intent of using them for
  other runtimes (e.g. kubernetes)
* split gardenruntime into its own package, separate from worker.Pool
* caches are no longer implemented as inputs, and cache paths are
  explicitly set in the ContainerSpec - it was easy to forget about
  them, and the original implementation deals with them in a very
  hacky/non-obvious way (if the cache isn't present on the worker,
  they're treated as remote inputs, so we create an empty volume for
  them, but they happen to not implement StreamableArtifactSource so we
  don't stream anything into them)
* code cleanup/method-call refactoring

behavioural
-----------
* resource result doesn't get saved on container properties, meaning
  re-attaching to a finished resource step won't give fetched versions
  (will fix in later commits)
* compression option is ignored, hardcoded to gzip (will fix in later
  commits)
* no p2p volume streaming (will implement in later commits)

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
allow creating containers/vols w/ fixed handles

I'm a bit hesitant to do this, because I intend to only use these new
methods in tests (at least for now), but this commit adds a method for
creating volumes with a fixed handle rather than a random UUID, and a
new ContainerOwner for referencing/creating Containers with a fixed
handle. it also allows finding creating volumes

the purpose of this is so that we can develop a fluent API for writing
tests for the new worker implementation that reference
containers/volumes with deterministic and semantic names.

while it makes the tests slower, I wanted to use the actual DB for these
worker tests so it's as close to real conditions as possible. I
considered writing hand-built fakes, but with the way the DB interfaces
are structured now, that would be a huge pain

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
add workertest/gardenruntimetest packages

these packages provide helpers for writing tests at a higher level than
just counterfeiter fakes

workertest provides a Scenario that has a reference to the DB (via
the dbtest package) and a list of workers, where a workertest.Worker
represents a builder for a single worker. it is currently implemented by
*gardenruntimetest.Worker, but will be implemented by a
k8sruntimetest.Worker in the future likely

gardenruntimetest provides in-memory implementations of a Baggageclaim
and Garden server, and a *Worker that can construct a real
*gardenruntime.Worker with the fake garden/baggageclaim

tests that use these test packages to follow in the next commit - just
splitting it up because these commits are all pretty huge

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
backport gardenruntime worker tests

these tests make use of the new gardenruntimetest and workertest
packages for higher level tests that should hopefully give us more
confidence than the existing atc/worker tests, and should be much more
resilient to refactors since the only interfaces we interact with are
the things we actually depend on, rather than interfaces we created just
for ease of testing. these tests are also just much more pleasant to
write IMO - it's cool to set up a scenario where you have some set of
workers, each with some set of containers and volumes, and exercise the
full system

the downside to these tests is that it's a lot harder to get high test
coverage for the error cases, but personally, I feel that the tradeoff
is worth it

also, worth noting that I wanted to use the stdlib testing package for
these tests in favour of Ginkgo, since I find the deeply nested
Context's that Ginkgo encourages result in tests that are ultimately
harder to read/write/modify, but I was "forced" to use Ginkgo since I
wanted to use the existing postgresrunner package which only currently
works with Ginkgo. so, I decided to write the tests as if I didn't have
BeforeEach/Context available to me, and personally I really like how
they turned out - the tests are all self-contained, so you don't need to
jump around the file to see why a test is failing.

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw commented:
FYI the intention is to take logic that previously existed in `worker.Client` and move it into the `exec` package - so, `exec.{Get,Check,Put,Task}Step` would interact directly with the runtime abstractions, which will hopefully be general enough that implementing a k8s or nomad runtime is as simple as implementing a (fairly) simple interface that is unaware of the business logic of these steps

aoldershaw committed:
S: use Compression instead of Encoding

I initially added runtime.Encoding to mirror Baggageclaim's interface,
but thinking about it more, it probably makes more sense to just pass in
a compression.Compression. it's a bit odd that compression.Compression
is aware of baggageclaim specifics, but it's not *that* specific - it's
just knowing that "gzip" is for gzip and "zstd" is for zstd

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
S: extract Streamer struct

that's responsible for handling streaming between volumes and streaming
in a file from a volume. this will be the target for re-implementing P2P
streaming

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: support P2P streaming

obviously it doesn't actually do anything yet since the worker2 package
still isn't wired up with the rest of the codebase. also extract some
common test setup logic to workertest

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: prefer local resource cache to remote

when an image/input wasn't fetched on the same worker as part of the
build, but there is an existing resource cache for the same volume, we
should use that instead. this behaviour already existed in
atc/worker/artifact_source.go

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: implement placement strategies

there's one behavioural difference currently - fewest-build-containers
previously fetched ActiveContainers from the containers table, rather
than the heartbeated value. I implemented it more simply using
db.Worker.ActiveContainers since #6469 will take care of this

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

evanchaoli commented:
Hey @aoldershaw, not sure if you have noticed #6495 that may impact this refactor.

aoldershaw commented:
@evanchaoli thanks, i've seen that PR, seems cool! I'm still not sure what will become of this PR, it's still very much a WIP - but if we decide to go forward with this PR, we'll be sure to bring your commits over

aoldershaw committed:
re-implement worker selection

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
S: expose gardenruntime.Container fields

to allow constructing the container in tests. I also realized we don't
need to pass in a db.Container, so I got rid of that, too

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
S: re-implement atc/resource

most of the logic for running resource scripts was previously encoded in
atc/worker/Worker.RunScript, which (unlike the name suggests) isn't for
running general scripts - it just worked for resources

since we're trying to make the actual runtime implementations lean,
runtime.Container just gives you a way to run generic things.
atc/resource is now responsible for interacting with a runtime.Container
to run resource scripts

this commit also gets rid of resource.Metadata, which was unused

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
S: return VolumeMounts in FindOrCreateContainer

...rather than placing that responsibility on the Container interface.
This is a bit nicer since it obviates the fact that
FindOrCreateContainer creates the necessary volumes too

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
S: replace runtime.Artifact with runtime.Volume

runtime.Artifact was being used for three things:

1. Task outputs (ID gives volume handle)
2. Get results (ID gives volume handle)
3. Task caches (ID gives some unique/deterministic string)

I previously eliminated the need for 3. by separating Caches from Inputs
on the ContainerSpec so that runtime implementations would have to
explicitly handle task Caches

1. and 2. are effectively the same thing - give me a way to identify a
volume. so, why not just use the volume itself? as a bonus, this means
we no longer have to unnecessarily lookup the volume for streaming since
we instead keep track of the in-memory representation of the Volume

note: this does not change how the exec package interacts with artifacts
in general, or fix tests - it just cleans up the build repository and
put inputs and gets rid of the interface.

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
S: remove runtime.VolumeSpec

I added this a bit preemptively, and the abstraction might not make
sense generally - it nearly identically mirrors baggageclaim.VolumeSpec,
and is only used internal to the gardenruntime package.

There will be a need to add a CreateVolume method to the runtime.Worker
interface for use in the artifactserver, but it's only ever used to
create empty volumes, so perhaps it's best to not try to generalize over
all volume providers (in case we end up going with something besides
baggageclaim for a particular runtime)

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: implement resource getter for running get steps

this re-implements worker.Fetcher and worker.FetchSource for running get
steps (under a lock) if necessary (i.e. the resource is not cached on
the worker). this should not change the behaviour of Fetcher/FetchSource

introducing bogus interfaces for the sake of easier testing is something
I tried to avoid in this refactor, but the exec tests are already a bit
of a mess and may require a large refactor of their own - until then,
I'm gonna test against this interface in get_step_test to keep my sanity

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: re-implement InitializeTaskCache

this commit re-implements atc/worker.Volume.InitializeTaskCache, which
is called in atc/exec/task_step.go. it also removes some of the test
setup helpers that effectively did the same thing for testing task
caches - now, the test hits the same code paths that the exec package
will, so the tests are closer to reality.

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
S: wire atc/exec with new worker package

this is mostly just a whole lot of plumbing - it doesn't actually change
the behaviour of anything. some notable changes:

* removed ArtifactSourcer entirely (input volumes are now referenced by
  handle and are looked up within the runtime implementation)
* use the new worker2.Streamer instead of worker.ArtifactStreamer
* emit starting event in the steps themselves (rather than requiring a
  StartingEventDelegate)
* task_step caches are set separately from inputs, corresponding with
  the change to the ContainerSpec

this doesn't touch the tests yet - those will come in a subsequent
commit

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: make Run/Attach API async

this now mirrors the Garden API more closely - Run/Attach give you a
runtime.Process, and you can .Wait on a runtime.Process to finish, which
gives you the runtime.ProcessResult

the reason for this change is to support modifying the process while its
running - in particular, `fly hijack` updates the TTY size when you
resize your terminal. so, the flow will be (in pseudo-code):

process := container.Run()
go func() {
    process.Wait()
    close(done)
}()
for {
    select {
    case <-done:
        return
    case <-stdin:
        ...
    case tty := <-resizeTerminal:
        process.SetTTY(tty)
    }
}

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: add SetTTY to runtime.Process

fly hijack requires setting the TTY whenever the terminal changes

this changes the behaviour from prior commits in that we only set the
large 500x500 window size for task steps. this matches the current
behaviour on master - for get/put/check, we just use the default

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: fix hijack API for new worker package

this required adding Env to the ProcessSpec - note that it's on the
ContainerSpec as well. not sure exactly how Garden deals with this -
presumably it appends the environment

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: fix artifactserver for new worker package

...and also fix a long-standing race condition on artifact volume
creation, as denoted by the big comment block in
artifactserver/create.go

...and we're finally compiling! ignoring the hundreds of tests i've
ignored

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: support absolute paths for task step

specifically, this commit adds support for absolute paths for:
* task run dir
* caches
* inputs
* outputs

unlike inputs/outputs, caches have to be passed in to the runtime as
relative (unless they're specified absolutely in the task config) for
when we lookup the task config by path

as with previous commits, tests to come

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: fix mount ordering

I introduced a bug where the input/output/cache bind mounts were defined
before the workdir bind mount, so when the workdir was mounted, it
cleared out all of the IO mounts. mounting the IO mounts after fixes
this problem

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: fix fly intercept fallback to sh

due to a brittle error check (not following wrapped errors), the fly
intercept fallback (when using the default shell) stopped working, as we
started wrapping the returned garden.ExecutableNotFoundError

since we want to avoid coupling ourselves to garden, this introduces a
new runtime.ExecutableNotFoundError, and runtime implementations must
return that error when the executable is not found

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
remove old worker package

...and move atc/worker2 -> atc/worker

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: use correct context for running steps

because we were using a non-standard name for contexts (processCtx), I
accidentally passed in the parent context (ctx) to running steps,
meaning the step timeouts weren't getting applied. this commit renames
processCtx -> ctx to make this less likely in the future

it also fixes a bug with the check step timeout where the global check
timeout no longer got applied if no timeout was set on the plan

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: use random process ID for checks

this way, when check containers are reused, we don't get errors
indicating that the process ID is already in use. we only got that error
with Guardian, as our containerd backend doesn't currently check for
this.

to accomplish this, Attach now takes in the process ID (rather than the
ProcessSpec) - task, get, and put all use hardcoded process IDs, whereas
check and hijack don't specify an ID (and so an ID is generated for
them)

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: add missing error check

was causing builds that were aborted to be marked as errored (with
"unexpected end of JSON" as the error)

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: keep streaming logs after process is aborted

I was passing in ctx to container.Run/container.Attach, which I learned
results in the logs after that point to not be streamed. so, when e.g. a
task is aborted, it'll be sent a SIGTERM, but if we cancel the ctx used
in container.Run, we won't see the logs of that process after that
point.

I decided to still pass in a ctx to runtime.Container.Run/Attach since
other implementations may be able to make use of the ctx more easily,
but the intention of that context is solely for the initial request. we
can evaluate whether or not passing in a context makes sense in the
future.

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

muntac commented:
@aoldershaw thanks for that great dependency graph! would it be easy enough to produce the same graph for the refactored code?

aoldershaw commented:
@muntac absolutely! I'm planning on writing out more in depth how things are structured, but hopefully this is a decent start:

![worker_dep_graph (1)](https://user-images.githubusercontent.com/26583442/111242281-3d6cc980-85d5-11eb-9142-fa9aebd7b7d4.png)

The primary goal is to find a common interface that can be implemented for other runtimes (e.g. a k8s runtime). One thing I didn't illustrate in the original dependency graph is how external packages interact with the `atc/worker` package, which was kind of a mixed bag - a ton of functionality was baked into the `atc/worker.Worker` interface, which wasn't really a general interface, having a lot of Garden/Baggageclaim specifics.

So, the `atc/runtime` package includes a new set of interfaces that can hopefully be more general - this consists of `Worker`, `Container`, `Volume`, and `Process` (not included in the diagram). `Worker` provides the ability to construct `Containers` and `Volumes`, and you can start `Processes` on `Containers` to run things.

The `atc/worker` package provides helpers for interacting with the runtime types. Specifically, there are two main components - `Pool` (which represents the pool of workers, allowing you to select a worker for running a container and locate `Volumes` and `Containers`), and `Streamer` (which understands how to stream data between `Volumes`, and stream files out of `Volumes`). Implementations of the `atc/runtime` types make use of `Pool` and `Streamer` to locate external volumes and stream data from them.

`atc/worker/gardenruntime` is an implementation of the `atc/runtime` interfaces that uses Garden for running containers and Baggageclaim for managing volumes. External packages aren't aware of any Garden/Baggageclaim specifics - they're only aware of `atc/runtime`

`atc/exec` (and `atc/resource`) implement the logic for all the step types (most relevant here are `task`, `get`, `put`, `check`, but also steps like `load_var` and `set_pipeline` which stream files out of volumes). In a simplified way, these steps select a `runtime.Worker` using `worker.Pool`, then create a `runtime.Container` on that worker, then run a `runtime.Process` on that container. The `GetStep` is a bit more complicated, and has a `resource.Getter` helper to achieve a similar thing).

muntac commented:
Still wrapping my head around this. Initially tried to account for the interactions of the `atc/worker` package with the rest of the `atc` i.e. what are the files outside the worker that rely on it and which functionality is being used. This is to ensure the refactor accounts for all of this, and to get a high level idea for what functionality the rest of the `atc` expects from the `atc/worker` package.

```
Files that use atc/worker:

(relative to atc/)
api/handler.go (for containerserver)
api/containerserver/hijack.go
api/containerserver/server.go
api/artifactserver/create.go
api/artifactserver/server.go

atccmd/command.go

engine/build_step_delegate.go
engine/builder.go
engine/check_delegate.go
engine/delegate_factory.go
engine/get_delegate.go
engine/put_delegate.go
engine/task_delegate.go

exec/artifact_input_step.go
exec/artifact_output_step.go
exec/build_step_delegate.go
exec/check_step.go
exec/get_step.go
exec/load_var_step.go
exec/put_step.go
exec/set_pipeline_step.go
exec/task_config_source.go
exec/task_step.go
```

Still figuring out the functionality bit, and will then dive back into commits. 
```
Current interface exposed by pool

pool
-> NewPool(WorkerProvider)
-> FindContainer(lager.Logger, int, string) (Container, bool, error)
-> FindVolume(lager.Logger, int, string) (Volume, bool, error)
-> CreateVolume(lager.Logger, VolumeSpec, WorkerSpec, db.VolumeType) (Volume, error)
-> ContainerInWorker(lager.Logger, db.ContainerOwner, WorkerSpec) (bool, error)
-> SelectWorker(context.Context, db.ContainerOwner,ContainerSpec,WorkerSpec,ContainerPlacementStrategy) (Client, error)
```
Feel free to suggest a different approach.

aoldershaw commented:
> Still figuring out the functionality bit

@muntac would it help if I went over each of the interactions with `atc/worker` in a bit more depth?

`api/containerserver/hijack.go` - this is the API handler that deals with running `fly hijack`/`fly intercept`, which allows you to run a process (typically a shell, but not necessarily) on an existing container. So, it interacts with `atc/worker` to locate a container, start a process, continually update the TTY (esp. terminal size), and store some state in the DB ("the container was hijacked at this time, so don't GC it")

---

`api/artifactserver/create.go` - this is the API handler used by `fly execute` to upload artifacts to a worker, where artifacts refers to local inputs that you upload to Concourse. It interacts with `atc/worker` to create an empty volume on a random worker and then stream in the artifact contents. This artifact (and its corresponding volume) can then be used to run steps (specifically, the `task` step that's run in `fly execute`). It is used by the `ArtifactInput` step

`exec/artifact_input_step.go` - this step runs in builds created by `fly execute`. It interacts with `atc/worker` to locate the volume that was previously created in `api/artifactserver/create.go` and makes it available to other steps.

`exec/artifact_output_step.go` - this step also runs in builds created by `fly execute` to mark a volume created by the `task` as a downloadable artifact. Previously, it interacted with `atc/worker` to locate the volume - now, it only interacts with the DB to mark a volume as an artifact **. Once an artifact has been "output" by this step, it can be downloaded in the api via `api/artifactserver/get.go`

** This is due to the refactor...previously, we used "Artifact" to mean two things - the uploadable/downloadable volumes I've been describing that are used in `fly execute`, but also as a general handle to a volume (formerly `runtime.Artifact`). If that's confusing, I'm not surprised - `runtime.Artifact` didn't make much sense to me to be honest, since it just seemed like an abstraction on top of volumes.

`api/artifactserver/get.go` - as I alluded to, this is the counterpart to `exec/artifact_output_step.go` - it's used by `fly execute` to download the `task`'s outputs  after it finishes. It interacts with `atc/worker` to locate the volume and stream the contents out.

So, the high-level flow for `fly execute` is (was):

```
fly execute -c task.yml -i some-input=./input -o some-output=./output
  -> POST api/artifactserver/create.go
    -> volume := atc/worker.Pool.CreateVolume() // for the artifact some-input
    -> volume.InitializeArtifact() // mark in the DB that this is an artifact
    -> volume.StreamIn(req.Body)
  -> Create a build with an ArtifactInputStep and ArtifactOutputStep (and a TaskStep of course)
    -> Run ArtifactInputStep
      -> atc/worker.Pool.LocateVolume() // find the artifact created earlier
      -> Register runtime.Artifact on build // basically just keep track of the fact that `some-input` points to that volume
    -> Run TaskStep
      -> Will give more details later, but this also keeps track of the output volume for `some-output`
    -> Run ArtifactOutputStep
      -> volume := atc/worker.Pool.LocateVolume() // find the volume that `some-output`, created in the task step, points to
      -> volume.InitializeArtifact() // mark in the DB that this is an artifact
  -> GET api/artifactserver/get.go
    -> volume := atc/worker.Pool.LocateVolume() // find the artifact volume created in ArtifactOutputStep
    -> volume.StreamOut() // download the contents via baggageclaim
```

---

will keep working on this description for the rest of the files. Let me know if this makes any sense/is helpful

muntac commented:
@aoldershaw saw this but forgot to respond. This is great. Thanks so much!

aoldershaw commented:
@muntac glad to hear it! still working on it, haven't gotten to the `exec` package interactions yet, but will update that comment soon

aoldershaw committed:
B: re-implement wait for worker (#6635)

the implementation is pretty much identical to #6635 with a couple
exceptions:

* Order for LimitActive{Containers,Volumes} now partitions the workers
  into two groups - those that satisfy the limit and those that don't.
  Those that don't are given lower priority, which should result in doing
  less work when Pick'ing a worker (since the workers that don't exceed
  the limit are attempted first)
* I got rid of Name() and NoWorkerFitContainerPlacementStrategyError
  because they were unused - since Order doesn't filter out workers,
  that error would never be returned

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: skip over non-running workers

I missed this condition when porting over the pool logic! whoops!

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: make unknown artifact source error generic

this takes an alternative approach to #6829, which missed the
set_pipeline step. making the error less specific seems reasonable
anyway, since the error is tied to a particular step

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
S: refactor runtimetest.{Container,Process}

the way it was setup before was a bit awkward and only checked the ID in
the spec, rather than the entire spec. by splitting up processes into
definitions and running processes, we can also catch issues like calling
Attach before Run, or processes not being called

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: ignore empty tags

this re-implements #6057. rather than adding logic to the worker pool,
this implementts this higher up - when we unmarshal tags (for both
worker registration AND on steps/resources/resource_types), we filter
out empty tags.

the reason for this is it seems better to handle this higher up so we
don't need to sprinkle checks for empty worker tags throughout the code.

since this behaviour is now further removed from where it's relevant,
though, I added an integration test to cover this case

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: fixup exec/engine unit tests

also made a few behavioural changes as a result of test failures:
* properly update resource config scope time on check script exiting
  with failed status code
* ensure a trailing slash is added to outputs (not sure why this is
  necessary - things seem to work without it - but there are tests for
  this explicitly so best not to mess with it)

some additional cleanup as well throughout

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
S: implement StreamIn/Out on runtimetest.Volume

and share the code with gardenruntimetest.Volume, since it's definitely
non-trivial

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
fixup API tests

this required some additions to the runtimetest types. specifically, we
needed thread-safety on accessing Processes to avoid tripping the race
detector when waiting for the hijacked process to come alive.

also, the StreamIn/Out behaviour now treats / as the same as ., since
there's some weirdness about using / in fstest.MapFS

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
fix data race in worker test

the race didn't actually affect the test since the reads are just for
polling, and we know that two writes can't occur simultaneously, so this
is just to shut the race detector up

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw commented:
@chenbh @muntac finally got around to fixing up the unit tests, and this PR is ready for review. Apologies that this PR got so huge - a lot of the code is “the same”, just moved around a lot, but it’s still a ton to review. I’ll finish writing up descriptions of interactions with the worker package soon, and am more than happy to go through the code with you

aoldershaw committed:
S: rename Pick -> Approve

porting over #6886

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
fix exec tests

post #6867

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: allow empty image spec

non-containerized platforms (darwin, windows) pass in an empty image
spec - this shouldn't result in an errored build

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: change attach error message

so that we can differentiate between Run and Attach failures

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
S: pass volumes instead of handles

previously, the runtime types that accepted volumes (Input.VolumeHandle,
ImageSpec.ImageVolume) took in the handles of volumes, and it was up to
the runtime implementations to lookup the volumes via the worker.Pool

however, since we store the full runtime.Volume in the
build.ArtifactRepository, we ended up doing an awkward dance where we
convert from Volume to Handle and then back to Volume

this was unnecessary and it complicated the implementation somewhat as
runtimes needed a reference to the worker pool solely to lookup volumes.

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
S: remove unnecessary var declaration

tiny refactor of something that was bothering me - better to declare the
variables closer to where they're used

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: remove unnecessary error

technically this is a behaviour change since it modifies the error
message, but that's the extent of it.
`ErrResourceConfigCheckSessionExpired` also has a similar (and arguably
less helpful) error message than `db.ContainerOwnerDisappearedError`,
and doesn't even mention anything about resource config check sessions -
just a generic error for all owner types.

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
S: add documentation for runtime types

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: stop logging version metadata in resource.Getter

it's super noisy as metadata can be quite long

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: backport caching streamed volumes

this commit backports #6660, which creates a worker_resource_cache on
the worker that a resource volume was streamed into. it does this by:

1. streaming the volume into an empty volume (like before). this volume
   becomes the resource cache (if the src volume was also a resource
   cache)
2. creating a COW volume over that empty volume and mounting that. this
   is to avoid mutating the cached volume.

that PR also introduces a change to the get step - if there is a cached
volume on any worker, use that, rather than re-running the get step.

some modifications were made over the original implementation:

* for "get from local cache", previously the get step would lookup all
  of the workers that contain the resource cache, and then walk through
  the list of workers and try to grab the volume from each one,
  stopping at the first success. now, more of this logic is pushed into
  the worker pool - the get step simply asked the pool for a resource
  cache volume (rather than the list of workers)

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
S: move gclient and transport under gardenruntime

it's garden/baggageclaim specific, so for now, let's move it under
gardenruntime

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

chenbh commented on atc/exec/build/repository.go:
```diff
@@ -21,37 +21,31 @@ type ArtifactName string
 // ArtifactName. Producers of artifacts, e.g. the Get step and the Task step,
 // will call this after they've successfully produced their artifact(s).
-func (repo *Repository) RegisterArtifact(name ArtifactName, artifact RegisterableArtifact) {
+func (repo *Repository) RegisterArtifact(name ArtifactName, volume runtime.Volume) {
```
I might be personally used to seeing `Artifact`, but I wonder if it's worth it to rename it to `Volume` everywhere. I think they're used interchangeably (with the exception of `api/artifactserver`) and might confuse people

aoldershaw replied:
Yeah fair point, it's a little awkward right now since it's only half committing to calling things volumes. It's weird having code like:

```go
volume, found := repository.ArtifactFor("foo")
```

On the flip side, we often describe inputs/outputs in a build as artifacts, so it'd be nice to keep that terminology.

WDYT about re-introducing a `runtime.Artifact` interface - only now, it's just the required subset of required methods (in particular, just `StreamOut`). Then, `runtime.Input` (and `runtime.ImageSpec`) could take a `runtime.Artifact` instead of a `runtime.Volume` - and runtime implementations could just special case on `Artifacts` that *also* happen to be volumes on the same worker.

(the motivation for something like this came from a discussion on discord where someone was looking for a feature to mount a volume to a `task` based on the contents of a variable - https://discord.com/channels/219899946617274369/413770960089382922/832341657202458744)

chenbh replied:
So the abstraction would be: an `Artifact` is an output of a step, you can grab the bits from it (`StreamOut`), but shouldn't be able to modify it. In this case an `Artifact` is passed from step to step and is purely a Concourse concept (i.e. it doesn't need to be generated from a container-based step). A `Volume` is the interface between `Artifact` and container-based steps, it has all the bells and whistles to take an `Artifact`, and convert it to a form that can be passed into a container runtime like garden.

I like it.

aoldershaw replied:
Just to make sure we're on the same page, my thinking was that a `Volume` would be a specific type of `Artifact` that is/was mounted to a container-based step. `Volume` aren't necessarily the *only* type of `Artifact`, though.

Any `Artifact` can be used as an input (or image), and it's up to the individual runtimes to convert `Artifacts` to mountable `Volumes`, if necessary. For the `gardenruntime`, this either mean streaming the contents of the `Artifact` into a new empty `Volume`, or creating a COW `Volume` if the `Artifact` was a `Volume` on the same worker.

When you create a container, you get back the list of `Volumes` that were mounted - these can then be registered to the build repository as `Artifacts`. So, when the `exec` package fetches an `Artifact` from the build repo, it has no idea whether it's a `Volume` or not - and it's up to the runtimes to figure out what to do with that (which may involve asking questions like "is this a `Volume`, and if so, what worker is it on?")

aoldershaw replied:
Take a look at https://github.com/concourse/concourse/pull/6597/commits/e4a4d2b99cb226cc518ba49d5ad7236741d4525f which implements this idea. I'm a bit worried it's premature abstraction, as it adds a *bit* of complexity to the runtime implementation and the `Streamer` (they now need to consider the case where the `Artifact` is not a `Volume`), but I don't think it's too bad. Anyway, happy to revert

chenbh replied:
+1 to most of what you said, the only thing I'm conflicted on is:

> When you create a container, you get back the list of `Volumes` that were mounted - these can then be registered to the build repository as `Artifacts`

Conceptually, `Artifacts` are the _outputs_ of a (previous) step, they should only be known at the end of a container's execution. But the way it's implemented is that it makes the underlying assumption that the volume mounts returned at the beginning of container creation are modified over the course of the container and will hold the output after the container has exited. Maybe `runtime.Container` interface should have a `OutputArtifacts()` that can be called after the process has exited to actually get the `Artifacts` that can be stored in the build repository.

But that's super pedantic and I'm fine with keeping the current implementation.

aoldershaw replied:
That's a good point. I took a look to remind myself how the `VolumeMounts` (returned by `FindOrCreateContainer`) are used, and it's used for 3 things:

1. Finding the resource volume for the `get` step to register as an artifact *and* to initialize the resource cache
2. Finding the output volumes of a `task` step to register as artifacts
3. Finding the cache volumes of a `task` step to initialize as task caches

For case 2, `OutputArtifacts` makes a lot of sense. For case 1, `OutputArtifacts` would have to still return `runtime.Volumes` (rather than `runtime.Artifacts). For case 3, we'd maybe need a `CacheVolumes` or something along those lines.

Cool if I just leave it as it is for now until we figure out a better approach?

chenbh reviewed (commented):
Some stuff caught by [staticcheck](https://staticcheck.io/)

```
atc/engine/task_delegate_test.go:148:6: func containerSpecDummy is unused (U1000)
atc/engine/task_delegate_test.go:169:6: func workerSpecDummy is unused (U1000)
atc/engine/task_delegate_test.go:177:6: func containerOwnerDummy is unused (U1000)
atc/worker/gardenruntime/gardenruntimetest/baggageclaim.go:78:2: ineffective assignment to field Baggageclaim.Volumes (SA4005)
atc/worker/gardenruntime/gclient/connection/connection_hijacker.go:74:12: httputil.NewClientConn has been deprecated since Go 1.0: Use the Client or Transport in package net/http instead.  (SA1019)
atc/worker/gardenruntime/gclient/connection/connection_test.go:346:12: limits.LimitInShares is deprecated: Use Weight instead.  (SA1019)
atc/worker/placement.go:16:40: duplicate struct tag "choice" (SA5008)
```

aoldershaw commented on atc/worker/gardenruntime/worker_test.go:
```diff
@@ -0,0 +1,1243 @@
+
+			Eventually(gardenContainer(container).NumProcesses).Should(Equal(1))
+
+			By("attaching to the running process", func() {
+				attachBuf := new(bytes.Buffer)
+				attachProcess, err := container.Attach(ctx, "process", runtime.ProcessIO{Stdout: attachBuf})
+				Expect(err).ToNot(HaveOccurred())
+
+				runResult, err := runProcess.Wait(ctx)
+				Expect(err).ToNot(HaveOccurred())
+				attachResult, err := attachProcess.Wait(ctx)
+				Expect(err).ToNot(HaveOccurred())
+
+				Expect(runResult.ExitStatus).To(Equal(0))
+				Expect(attachResult.ExitStatus).To(Equal(0))
+
+				Expect(runBuf.String()).To(Equal("hello world\n"))
+				Expect(attachBuf.String()).To(Equal("hello world\n"))
+			})
```
Not sure I understand what you mean - isn't that what this is testing?

chenbh replied:
Whoops, should've clarified that I meant re-attaching after a re-attach. This tests that you can re-attach after a run, but what about a second re-attach?

aoldershaw replied:
Modified the test to `Attach` twice

aoldershaw commented on atc/api/artifactserver/create.go:
```diff
@@ -20,47 +20,20 @@ func (s *Server) CreateArtifact(team db.Team) http.Handler {
 			return
 		}
 
-		// NOTE: there's a race condition here between when the
-		// volume gets created and when the artifact gets initialized
-
-		// Within this timeframe there's a chance that the volume could
-		// get garbage collected out from under us.
-
-		// This happens because CreateVolume returns a 'created' instead
-		// of 'creating' volume.
-
-		// In the long run CreateVolume should probably return a 'creating'
-		// volume, but there are other changes needed in FindOrCreateContainer
-		// with the way we create volumes for a container
-
-		// I think leaving the race condition is fine for now. Worst case
-		// is a fly execute will fail and the user will need to rerun it.
```
I don't think so, because (at least based on this comment block) creating volumes don't get GC'd, and we only mark the volume as created after it's been initialized

chenbh replied:
Ah, missed the `CreateVolume` returning a `creatingVolume` part

aoldershaw commented on atc/db/volume_repository.go:
```diff
@@ -451,19 +465,15 @@ func (repository *volumeRepository) FindResourceCacheVolume(workerName string, r
 	return createdVolume, true, nil
 }
 
-func (repository *volumeRepository) FindCreatedVolume(handle string) (CreatedVolume, bool, error) {
-	_, createdVolume, err := getVolume(repository.conn, map[string]interface{}{
+func (repository *volumeRepository) FindVolume(handle string) (CreatingVolume, CreatedVolume, error) {
```
Yeah no real reason - guessing I needed it at one point but then changed my approach. Will revert

aoldershaw commented on atc/worker/pool.go:
```diff
@@ -10,143 +10,112 @@ import (
-type NoCompatibleWorkersError struct {
-	Spec WorkerSpec
-}
+type Pool struct {
+	factory       Factory
+	db            DB
+	workerVersion version.Version
```
That seems pretty reasonable to me. I wonder what the reason is for using the random placement strategy for checks - I had forgotten about that/hadn't thought too much about it at the time. It seems weird because then we'd be ignoring the e.g. `limit-active-containers/volumes` strategies and could knowingly place checks on a "bad" worker

aoldershaw commented on atc/atccmd/command.go:
```diff
@@ -1024,46 +995,18 @@ func (cmd *RunCommand) backendComponents(
-	dbTaskCacheFactory := db.NewTaskCacheFactory(dbConn)
-	dbWorkerTaskCacheFactory := db.NewWorkerTaskCacheFactory(dbConn)
 	dbVolumeRepository := db.NewVolumeRepository(dbConn)
+	resourceGetter := resource.NewGetter(lockFactory, clock.NewClock(), dbResourceCacheFactory, dbVolumeRepository)
```
Agreed that `resource.Getter` doesn't really have a strong identity, but not sure I agree with turning it into a `ResourceFactory`/`ContainerFactory`. `resource.Getter` is really just responsible for executing a part of the get step - specifically, looking up the resource cache, and if it doesn't exist, running the `in` script on resources under a lock and storing it as a cache. It doesn't know about how to create the container nor how to create the resource (which is just data). It also doesn't *really* know how to run the `in` script, and can easily be modified to support prototype-based resources by just taking an interface rather than the concrete `resource.Resource`.

IMO the better approach would be to move this logic into either the `exec.GetStep` itself or `engine.GetStepDelegate`, or some combination thereof (letting the delegate handle the DB interaction). The main reason for not doing this initially was so that I could test this logic in isolation, but it may not actually be that bad testing at the level of the get step.

aoldershaw replied:
Take a look at https://github.com/concourse/concourse/pull/6597/commits/ab30714be27030b68b857b62740066fa51f316ca, which implements this idea. Moving the `resource.Getter` into `exec.GetStep` also allowed for simplifying some redundant cache-checking logic that was added for caching streamed volumes.

p.s. sorry for adding yet another big commit into this already massive PR

aoldershaw committed:
address linter warnings

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: move resource.Getter into exec.GetStep

resource.Getter never really made sense as an abstraction. initially, I
added it to make it easier to write the get step tests since the logic
is a bit awkward, but found it was actually easier to not fake it out.

given that, there's no real point of resource.Getter, since what it does
it really just the implementation of the get step. so, this commit moves
the responsibility of resource.Getter into exec.GetStep.

there are a few behavioural differences to be aware of:

* When caching streamed resources is enabled, there is now a single lock
  for the particular resource cache (rather than one per worker). This
  means that if the cache doesn't exist yet, and multiple get steps are
  run concurrently for the same resource + version on different workers,
  then only one will run the get script. before, it was possible for
  both to run concurrently.
* Lock name is now determined by the resource cache ID rather than the
  source + params + version, since the latter may not uniquely identify
  the resource (since it doesn't take into account the base resource
  type). I think this was a bug.
* We no longer lookup the cache after acquiring the lock. While it's
  technically possible that we will run the get script unnecessarily
  (since the resource could get cached between looking up the cache
  initially and acquiring the lock), we have concurrency controls when
  initializing resource caches to ensure a single resource cache volume
  will be created. This extra lookup adds complexity and has a performance
  impact, and it should be a rare case.
* The step timeout no longer takes into account the time waiting for the
  lock

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
fix compile error from rebasing onto prototypes

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
S: revert VolumeRepository.FindVolume signature

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
test attaching multiple times

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
S: split up runtime.Artifact from runtime.Volume

this commit adds a new abstraction, runtime.Artifact, that is simply
some chunk of data that can be streamed out. runtime.Artifact is a
generalization of runtime.Volume. the exec package only interacts with
runtime.Artifact - it doesn't care whether the Artifact is from a Volume
or anything else.

note that we don't currently have any types of Artifact that aren't
Volumes - however, this means we could support mounting volumes to
workers from content that is generated on the ATC in memory (rather than
needing a dummy task that creates an output with the file contents,
which is much more heavy weight).

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw commented on atc/worker/gardenruntime/worker_test.go:
```diff
@@ -0,0 +1,1243 @@
+	ctx := context.Background()
+
+	Test("running a process on a newly created container", func() {
+		scenario := Setup(
+			workertest.WithWorkers(
+				grt.NewWorker("worker"),
+			),
+		)
+		worker := scenario.Worker("worker")
+
+		container, volumeMounts, err := worker.FindOrCreateContainer(
+			ctx,
+			db.NewFixedHandleContainerOwner("my-handle"),
+			db.ContainerMetadata{},
+			runtime.ContainerSpec{
+				Dir: "/workdir",
+				ImageSpec: runtime.ImageSpec{
+					ImageURL: "raw:///img/rootfs",
+				},
+			},
+		)
+		Expect(err).ToNot(HaveOccurred())
+
+		By("running a process on that container", func() {
+			buf := new(bytes.Buffer)
+			process, err := container.Run(
+				ctx,
+				runtime.ProcessSpec{
+					Path: "echo",
+					Args: []string{"hello world"},
+				},
+				runtime.ProcessIO{
+					Stdout: buf,
+				},
+			)
```
Honestly I initially wanted to just use a real garden server backed by houdini and a real baggageclaim server that just writes to a temp dir, but found that it was easier to just write out fake them out (and likely less flaky, too).

I did want it to be as close as (realistically) possible to the real thing. For the most part, I wanted the `gardenruntimetest` to insulate the tests from the garden/baggageclaim specifics so as to *not matter* whether you were talking to a real garden or baggageclaim when reading/writing the tests. Of course, that wasn't totally achieved (e.g. special casing on the `Path` of the `ProcessSpec` to perform different things).

That's good feedback that it was hard to reason about, though. I'll try to think about how to make it more clear, but would love to hear any suggestions if you have them

chenbh reviewed (commented):
I think I'm reaching the limit of being able to grasp the code in my head. There's no obvious logic errors as far as I can tell, I'm down to just deploy this on prod and monitor it for a week

aoldershaw committed:
adjust random placement strategy comment

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw commented on atc/resource/resource_test.go:
```diff
@@ -1,50 +1,339 @@
-package resource_test
+package resource
```
I think it was more because there wasn't an obvious benefit to using a `_test` package here. The [CodeReviewComments](https://github.com/golang/go/wiki/CodeReviewComments#import-dot) guidelines mentions, when discussing `import ( . package )`:

> The import . form can be useful in tests that, due to circular dependencies, cannot be made part of the package being tested

...which implies that `*_test` packages aren't just about testing the package via only its public interface - it's also (more?) about avoiding import cycles.

Looking at some large Go projects (including the Go standard library), from what I've seen, most of the time tests reside in the same package as the impl: e.g. [Nomad](https://github.com/hashicorp/nomad/blob/main/scheduler/annotate_test.go), [Kubernetes](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/daemon_controller_test.go), [Go standard library](https://github.com/golang/go/blob/master/src/net/addrselect_test.go).

That said, if you feel strongly, I don't mind changing it

aoldershaw committed:
B: use DB container counts rather than reported

...for the fewest-build-containers strategy. now that this is cached, it
should be very quick to access

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
fix tsa test package path

concourse/concourse#7336 sped up the TSA tests by compiling the test
binary with a package constant set. however, this package was moved, so
we weren't setting the constant properly. unfortunately, this doesn't
result in a compilation error, and fails silently

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
Merge branch 'master' into rewrite-atc-worker
